File: application/src/main/java/application/ElevatorService.java
package application;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.List;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service
public class ElevatorService {
    private static final Logger LOG = LoggerFactory.getLogger(ElevatorService.class);

    private static final List EMPTY_LIST = Collections.emptyList();

    private final ElevatorRepository repository;
    private final Lock mutateHailsLock = new ReentrantLock();

    @Autowired
    public ElevatorService(ElevatorRepository repository) {
        this.repository = repository;
    }

    public synchronized List<RequestHail> getNextHails(final double yCoordinate,
                                          final Direction elevatorDirection) {


        final Deque<RequestHail> hails = repository.getHails();

        if (hails.isEmpty()) {
            return EMPTY_LIST;
        }

        final Direction directionToUse;

        if (elevatorDirection == Direction.IDLE) {
            directionToUse = hails.peek().getDirection(); // pick a direction to go
        } else {
            directionToUse = elevatorDirection;
        }

        Stream<RequestHail> filteredHails = hails.stream()
                .filter(hail -> directionToUse == hail.getDirection());

        if (elevatorDirection != Direction.IDLE) {
            filteredHails = filteredHails.filter(hail -> {
                double hailFloorNumber = (double) hail.getFromFloorNumber();
                return directionToUse == Direction.UP
                        ? yCoordinate <= hailFloorNumber
                        : yCoordinate >= hailFloorNumber;
            });
        }

        final List<RequestHail> requestHails = filteredHails.collect(Collectors.toUnmodifiableList());

        hails.removeAll(requestHails);
        return requestHails;

    }

    public List<Integer> getAndClearNextFloorDestinationForElevatorId(int elevatorId) {
        List<Integer> nextDestinations = repository.getFloorDestinationsForElevatorId(elevatorId);
        if (nextDestinations.isEmpty()) {
            return EMPTY_LIST;
        }
        List<Integer> listForElevator = new ArrayList<>(nextDestinations); // shallow copy but with primitives
        nextDestinations.clear(); //#check repository too

        return listForElevator;
    }

}

File: application/src/main/java/application/ElevatorService.kt
package application

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.*
import java.util.stream.Collectors

@Service
class ElevatorService @Autowired constructor(private val repository: ElevatorRepository) {
    companion object {
        private val LOG: Logger = LoggerFactory.getLogger(ElevatorService::class.java)
    }

    @Synchronized
    fun getNextHails(yCoordinate: Double,
                     elevatorDirection: Direction): List<RequestHail> {
        val hails = repository.hails
        if (hails.isEmpty()) {
            return emptyList()
        }

        val directionToUse: Direction = if (elevatorDirection === Direction.IDLE) {
            hails.peek().direction!! // pick a direction to go
        } else {
            elevatorDirection
        }

        var filteredHailsStream = hails.stream()
                .filter { hail: RequestHail -> directionToUse === hail.direction }
        if (elevatorDirection !== Direction.IDLE) {
            filteredHailsStream = filteredHailsStream.filter { hail: RequestHail ->
                val hailFloorNumber = hail.fromFloorNumber.toDouble()
                if (directionToUse === Direction.UP) yCoordinate <= hailFloorNumber else yCoordinate >= hailFloorNumber
            }
        }
        val requestHails = filteredHailsStream.collect(Collectors.toUnmodifiableList())
        hails.removeAll(requestHails.toSet())
        return requestHails
    }

    fun getAndClearNextFloorDestinationForElevatorId(elevatorId: Int): List<Int> {
        val nextDestinations = repository.getFloorDestinationsForElevatorId(elevatorId)
        if (nextDestinations.isEmpty()) {
            return emptyList()
        }
        val listForElevator: List<Int> = ArrayList(nextDestinations) // shallow copy but with primitives
        nextDestinations.clear() //TODO change to atomic compare and set
        return listForElevator
    }


}